resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java

```diff
@@ -51,25 +51,40 @@ public void work(ServletOutputStream sos) {
       private final byte[] bytes;
       private final int offset;
       private final int length;
+      private final Queue<AsyncOperation> asyncQueue;
+      private boolean dataWritten;
 
-      public WriteOperation(final OutputStream stream, final byte[] bytes, final int offset, final int length)
+      public WriteOperation(final OutputStream stream, final byte[] bytes, final int offset, final int length,
+                            final Queue<AsyncOperation> asyncQueue)
       {
          super(stream);
          this.bytes = bytes;
          this.offset = offset;
          this.length = length;
+         this.asyncQueue = asyncQueue;
       }
 
       @Override
       protected void doWork(ServletOutputStream sos)
       {
          try
          {
+            if(dataWritten) {
+               future.complete(null);
+               return;
+            }
             // we only are complete if isReady says we're good to write, otherwise
             // we will be complete in the next onWritePossible or onError
             if(sos == null || sos.isReady()) {
                stream.write(bytes, offset, length);
-               future.complete(null);
+               dataWritten = true;
+               if(sos == null || sos.isReady()) {
+                  future.complete(null);
+               } else {
+                  // Queueing this operation to complete the future later
+                  // when the stream is ready.
+                  asyncQueue.add(this);
```

Comment:
The one problem I see here is that the queue is added to out of its lock.

In a lot of ways I really just want to re-write the whole deferred stream, but I'm not sure if I should or not :)

Reply:
It is effectively locked as we call the work only from [here](https://github.com/resteasy/resteasy/blob/6f123411d72e4fe5b5b38fbc54cbb3b114f528fe/resteasy-core/src/main/java/org/jboss/resteasy/plugins/server/servlet/HttpServletResponseWrapper.java#L271) which is inside a synchronized block, right?

Reply:
Ah yeah, it would be you're right. Sorry about that :)

Reply:
I did file https://github.com/resteasy/resteasy/pull/3048 as an example too. I think it needs a few changes, but it seems to work as well.